Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -383,70 +392,59 @@ protected void OnStalling(ChainedBlock chainedBlock)
         /// <param name="vectors">Information about blocks to download.</param>
         /// <param name="innernodes">Available nodes to distribute download tasks among.</param>
         /// <param name="minHeight">Minimum height of the chain that the target nodes has to have in order to be asked for one or more of the block to be downloaded from them.</param>
-        /// <remarks>TODO: This function is suspected to suffer from https://github.com/stratisproject/StratisBitcoinFullNode/issues/233 .</remarks>
-        private void DistributeDownload(InventoryVector[] vectors, BlockPullerBehavior[] innernodes, int minHeight)
+        private void DistributeDownload(Dictionary<int, InventoryVector> vectors, BlockPullerBehavior[] innernodes, int minHeight)
         {
-            if (vectors.Length == 0)
+            if (vectors.Count == 0)
                 return;
 
+            List<int> requestedBlockHeights = new List<int>();
+            List<DownloadAssignmentStrategy.PeerInformation> peerInformation = new List<DownloadAssignmentStrategy.PeerInformation>();
 
-            // Be careful not to ask the block from a node that does not have it 
-            // (we can check the ChainHeadersBehavior.PendingTip to know where the node is standing).
-            var selectnodes = new List<BlockPullerBehavior>();
             foreach (BlockPullerBehavior behavior in innernodes)
             {
-                // filter nodes that are still behind using the 
-                // pending tip in the chain behaviour
                 if (behavior.ChainHeadersBehavior?.PendingTip?.Height >= minHeight)
-                    selectnodes.Add(behavior);
+                {
+                    DownloadAssignmentStrategy.PeerInformation peerInfo = new DownloadAssignmentStrategy.PeerInformation()
+                    {
+                        QualityScore = behavior.QualityScore,
+                        PeerId = behavior,
+                        ChainHeight = behavior.ChainHeadersBehavior.PendingTip.Height
+                    };
+                    peerInformation.Add(peerInfo);
+                }
             }
-            innernodes = selectnodes.ToArray();
 
-            if (innernodes.Length == 0)
+            Dictionary<DownloadAssignmentStrategy.PeerInformation, List<int>> assignBlocksToPeers = DownloadAssignmentStrategy.AssignBlocksToPeers(requestedBlockHeights, peerInformation);
```

Comment:
probably better if this is called blocksAssignedToPeers. The following foreach is then clearer to read.